thunder/tests/opinfos.py

```diff
@@ -3609,6 +3609,47 @@ def reshape_sample_generator(op, device, dtype, requires_grad, **kwargs):
 shape_ops.append(reshape_opinfo)
 
 
+def unflatten_sample_generator(op, device, dtype, requires_grad, **kwargs):
+    make = partial(make_tensor, device=device, dtype=dtype, requires_grad=requires_grad)
+
+    cases = (
+        ((4,), 0, (4,)),  # no-op
+        ((2, 2), -1, (2,)),  # no-op
+        ((1, 2, 1), 1, (2,)),  # no-op
+        ((4, 2), 0, (2, 2)),
+        ((125, 3), 0, (25, 5)),
+        ((25, 25), -1, (1, 5, 5)),
+        ((16, 32), -1, (2, 4, 1, 4)),
+        ((4, 5, 6), -1, (6, 1, 1, 1)),
+    )
+
+    for tensor_shape, dim, shape in cases:
+        yield SampleInput(make(tensor_shape), dim, shape)
+
+
+def unflatten_error_generator(op, device, dtype=torch.float32, **kwargs):
+    make = partial(make_tensor, dtype=dtype, device=device)
+    input_tensor = make(4, 4)
+    dim = 3
+    yield (SampleInput(input_tensor, 0, ()), RuntimeError, r"unflatten\(\) sizes must be non-empty")
```

Comment:
Two more error cases that are interesting:

- the unflattened shape is fully specified and invalid. For example, a dimension of length 12 and a request to unflatten it to `(4, 4)`
- the unflattened shape is partially specified and invalid. For example, a dimension of length 12 and a request to unflatten it to `(-1, 5)`.

I imagine that the call to `view()` will handle these errors properly for these cases, but it's still nice to check them from `unflatten()`

Reply:
Awesome idea! I thought error cases should only include the ones caught by this specific function (unflatten) but it makes sense to check these.